libraries/utils/src/render-lines-on-sp.ts

```diff
@@ -63,77 +61,69 @@ export class ScatterPlotLinesRenderer {
+                    this.paths[i] = line;
                     line.moveTo(aX!, aY!);
-                    const color = lineOpts.color ?? '0,128,0';
-                    const opacity = lineOpts.opacity ?? 1;
-                    ctx.strokeStyle = `rgba(${color},${opacity})`;
-                    ctx.lineWidth = lineOpts.width ? lineOpts.id === this.currentLineIdx ? lineOpts.width + 2 : lineOpts.width :
-                        lineOpts.id === this.currentLineIdx ? 3 : 1;
+                    const color = this.lines.colors?.[i] ? this.lines.colors?.[i] : '0,128,0';
+                    const opacity = this.lines.opacities?.[i] ? this.lines.opacities?.[i] : 1;
+                    this.ctx.strokeStyle = `rgba(${color},${opacity})`;
+                    this.ctx.lineWidth = this.lines.widths?.[i] ? i === this.currentLineIdx ? this.lines.widths?.[i] + 2 :
+                    this.lines.widths?.[i] : i === this.currentLineIdx ? 3 : 1;
                     let midPoint: DG.Point = this.midPoint(pointFrom, pointTo);
                     if (this.numberOfLinesPerPair[pairStr] > 1 || linesPerPairCounter[pairStr] > 1) {
                         midPoint = this.findPerpendicularPointOnCurve(linesPerPairCounter[pairStr],
                             pointFrom.x, pointFrom.y, midPoint.x, midPoint.y)
                     }
                     line.quadraticCurveTo(midPoint.x, midPoint.y, bX, bY);
-                    if (lineOpts.directionCol) {
-                        const arrowPoint = this.getArrowPoint(lineOpts, lineOpts.points[0], lineOpts.points[1], pointFrom, pointTo);
-                        if (arrowPoint)
-                            this.canvasArrow(line, arrowPoint.x, arrowPoint.y, midPoint.x, midPoint.y);
-                    }
-                    ctx.beginPath();
-                    ctx.stroke(line);
-                    ctx.closePath();
+                    if (this.lines.drawArrows?.getBit(i))
+                        this.canvasArrow(line, bX, bY, midPoint.x, midPoint.y);
+                    this.ctx.beginPath();
+                    this.ctx.stroke(line);
+                    this.ctx.closePath();
                 }
             }
         }
-
     }
 
-    countLinesPerPairAndSort(): void {
-        for (let i = 0; i < this.lines.length; i++) {
-            const pair = this.lines[i].points;
-            const p1 = pair[0] > pair[1] ? pair[1] : pair[0];
-            const p2 = pair[0] > pair[1] ? pair[0] : pair[1];
-            pair[0] = p1;
-            pair[1] = p2;
-            const pairStr = `${this.lines[i].points[0]}#${this.lines[i].points[1]}`;
-            if (!this.numberOfLinesPerPair[pairStr])
-                this.numberOfLinesPerPair[pairStr] = 1;
-            else
-                this.numberOfLinesPerPair[pairStr]++;
+    countLinesPerPair(): void {
+        for (let i = 0; i < this.lines.from.length; i++) {
+            const pairStr = this.getStringPair(this.lines.from[i], this.lines.to[i]);
```

Comment:
StringPair is inefficient, come up with a better solution. One possibility is to sort lines, this way we can count them very efficiently

Reply:
Sorting is not that trivial in this case. At least because the arrow direction can be set through points indexes order. For example we have [1, 4] in 'from' array and [4, 1] in 'to' array. But those are for the same point pair. At the moment I start to render I need to know the following for each point pair:
1) are there multiple lines for this pair of point (in this case I start drawing the first line as a curve) to form smth like this
![image](https://github.com/datagrok-ai/public/assets/44060410/da6fe246-c726-4db1-9608-3b42e3f9cef2)
2) the index of the line for this particular pair (to calculate the height of the curve and the side where it will be drawn).

I have proposed a decision without string pairs - createMultiLinesIndices() method which is only called once in the constructor. And generates an array with required information (basically a Uint32Array with an index of the line in case there are multiple lines for the pair) @skalkin , please take a look.

Reply:
Thanks. So we are creating at least an additional string for each line? I don't like it, there are more efficient ways to do that (for instance by using a bit array to indicate whether we already have lines connecting the points, and then keeping a list of such lines somewhere).